rgsync/Connectors/redis_connector.py

```diff
@@ -56,81 +100,108 @@ def WriteData(self, data):
             WriteBehindLog("Warning: in connector received empty batch to write")
             return
 
+        if not self.session:
+            self.session = self.connection.Connect()
+        
+        # in case of exactly once get last id written
+        shardId = None
         try:
-            if not self.session:
-                self.session = self.connection.Connect()
-                if self.exactlyOnceTableName is not None:
-                    shardId = 'shard-%s' % hashtag()
-                    res = self.session.execute_command('HGET', self.exactlyOnceTableName, shardId)
-                    if 'ERR' in str(res):
-                        WriteBehindLog("Error: Failed to get entry for exactly once key: %s, field: %s" 
-                                % (self.exactlyOnceTableName, shardId))
-                        WriteBehindLog("Error: Command HGET execution failed, error = %s" % str(res))
-                        raise Exception(res) from None
-                    elif res is not None:
-                        self.exactlyOnceLastId = str(res)
-                    else:
-                        self.shouldCompareId = False
+            if self.exactlyOnceTableName is not None:
+                shardId = 'shard-%s' % hashtag()
+                res = self.session.execute_command('HGET', self.exactlyOnceTableName, shardId)
+                if res is not None:
+                    self.exactlyOnceLastId = str(res)
+                else:
+                    self.shouldCompareId = False
 
         except Exception as e:
-            self.session = None     #Next time we will connect to database
+            self.session = None             #Next time we will connect to database
             self.exactlyOnceLastId = None
             self.shouldCompareId = True if self.exactlyOnceTableName is not None else False
-            msg = "Failed to connect to Redis database, error = %s" % str(e)
+            msg = "Exception occur while getting shard id for exactly once. Exception : {}".format(e)
             WriteBehindLog(msg)
             raise Exception(msg) from None
 
-
+        pipe = self.session.pipeline()
```

Comment:
here I think you need "transaction=True" if you have exactly once table otherwise "transaction=False" also if the target redis is a cluster then you can not support transaction (maybe rais an exception?)

Reply:
@MeirShpilraien 
I am not getting what exactly you are suggesting regarding cluster and transaction logic.
Please elaborate with an example.

Reply:
@ratnesh-tiwari so basically for the exactly once property you have have the update inside transaction but Redis cluster do not support transaction over multiple shards so basically you can not easily support it. 

I think the best way it to allow, when creating the connection object, to specify if the target is cluster or single shard. If its a cluster and also the exactly once property is requested then we should raise an exception that will reach the user as an error indicating that this combination is not supported.